_posts/2019-04-10-imperatywnie-vs-funkcyjnie.md

```diff
@@ -0,0 +1,136 @@
+
+```kotlin
+fun multiply(val a: Int, val b: Int) = a * b
+fun divide(a: Int, b: Int) = a / b
+ 
+multiply(123456789, 123456789) // -1757895751
+divive(123456789, 0) // ArithmeticException: / by zero
+```
+Powyższy program nie jest funkcyjny dlatego, że mogą w nim wystąpić efekty uboczne. Jeśli przekręcimy Inta to dostaniemy ujemną wartość - błędną wartość. W drugim przypadku dostaniemy wyjątek, czyli efekt uboczny czego również się nie spodziwaliśmy. Nasze początkowe założenie, że funkcja pomnoży/podzieli wynik jest błędna. Oczywiście rzadko kiedy sytuacja jest, aż tak trywialna, ale można sobie wyobrazić, że to wszystko jest jakimś zapytaniem do bazy, albo inną operacją. Chodzi o to, aby nasze funkcje były deterministyczne, czyli wynik zawsze jest taki sam dla podanych argumentów. Jeśli są jakieś efekty uboczne to wiemy, jakie i odpowiednio na nie reagujemy. 
+
+```kotlin
+fun multiply(a: Int, b: Int) = a * b.toFloat()
+fun divide(a: Int, b: Int) = a / b.toFloat()
+ 
+multiply(123456789, 123456789) // 1.52415794E16
+divide(123456789, 0) // Infinity
+
+result.isInfinite().maybe { print("You should learn the basics of math, probably.") }
+```
+
+W tym momencie dla każdej możliwej wartości będzie ten sam wynik. Również w przypadku poprzedniego wyjątku teraz 
+zostanie zwrócony obiekt **Infinity**, na którym możemy działać dalej z naszym kodem. Można użyć extension function z Kotlina, które pozwala rozszerzyć każdą metodę o dodatkowe funkcjonalności. W tym przypadku zaciągnęliśmy biblioteką Arrow, która zrobiła za nas trochę magii `maybe { ... }`, które zwraca `none()` dla braku wartości lub `some()` jeśli coś jest.
+
+Gdy wiemy czego się spodziewać po funkcji i zawsze zwraca ona prawidłową wartość można nazwać ją czystą. Znaczy to tyle, że takie pure functions nie mają żadnych efektów ubocznych. To co zwraca zależy tylko i wyłącznie od parametrów jakie podaliśmy. Dobrym przykładem są funkcje z klasy Math. Gdzie biorąc `sqrt(2.0)` wiemy, że wynikiem jest zawsze jakiś obiekt typu Double (jak było wcześniej powiedziane w Kotlinie wszystko jest obiektem również prymitywy, których bądź co bądź po prostu nie ma).
+
+## To jak pisać bardziej funkcyjnie?
+
+```kotlin
+// Na początek imperatywny przykład
+fun buyBook(bookName: String, creditCard: CreditCard): Book {
+    val book = Book(name = bookName)
+    creditCard.performPayment(book.price)
+    return book
+}
+```
+
+Powyższy kod jest trudny do przetestowania. Musielibyśmy sobie zamockować proces weryfikacji w banku, czy karta w ogóle istnieje, czy posiada środki na koncie i tak dalej. Dopiero po tym zwracamy film. Zróbmy to trochę inaczej. Co jeśliby przetestować to bez kontaktu z bankiem? Moglibyśmy powiązać płatność z książką.
+
+```kotlin
+// Bardziej funkcyjny przykład
+fun buyBook(bookName: String, creditCard: CreditCard): Pair<Book, Payment> {
```

Comment:
Zastanowiłbym się czy na tym etapie warto zwracać uwagę na Pair, Tuple etc - nie wiem czy dla początkujących czytelniejszym nie byłby obiekt typu Purchase(val book:Book, val payment:Payment). 
Dodatkowo, jeśli byśmy coś takiego zrobili to można by oprzeć się na tym w kolejnym artykuliku, który mógłby dotyczyć kompozycji.

Reply:
Zmieniłem na `Purchase` oraz poprawiłem trochę kodu, bo były tam `class`, a powinny być `data class`.